libcaf_core/caf/detail/type_id_list_builder.cpp

```diff
@@ -102,7 +102,7 @@ void type_id_list_builder::reserve(size_t num_elements) {
 }
 
 void type_id_list_builder::push_back(type_id_t id) {
-  if ((size_ + 1) >= reserved_)
+  if (size_ >= reserved_)
```

Comment:
After reading this a couple of times and checking with `reserve`, I'm not sure whether this patch is correct.

This is the start of `reserve`:

```c++
void type_id_list_builder::reserve(size_t num_elements) {
  auto new_capacity = num_elements + 1; // One extra slot for the size prefix.
  if (reserved_ >= new_capacity)
    return;
  reserved_ = new_capacity;
  auto ptr = realloc(storage_, reserved_ * sizeof(type_id_t));
  ...
```

So `reserved_` is basically `max_size + 1`. I think that's the reason why this code checks for `if ((size_ + 1) >= reserved_)`: because `size_` is the number of elements (does not contain the dummy) but `reserved_` is the number of elements *plus the dummy*. I do agree that it's a bit confusing. If we'd go with your suggestion, we would also need to change how `reserved_` works (by setting it to `num_elements` instead of `new_capacity`). Something like this:

```c++
void type_id_list_builder::reserve(size_t num_elements) {
  if (reserved_ >= num_elements)
    return;
  reserved_ = num_elements;
  auto new_capacity = num_elements + 1; // One extra slot for the size prefix.
  auto ptr = realloc(storage_, new_capacity * sizeof(type_id_t));
  ...
```

But feel free to just drop this change from the commits instead and I'll clean that up at some later point.

Reply:
Sorry, discard this comment. `size_` also includes the dummy.

Reply:
I think `size_` contains the index for next element, because `size()` returns `size_ - 1`;